src/caffe/layers/conv_layer.cpp

```diff
@@ -74,10 +74,76 @@ void ConvolutionLayer<Dtype>::Backward_cpu(const vector<Blob<Dtype>*>& top,
   }
 }
 
-#ifdef CPU_ONLY
+#if defined(USE_OPENCL)
+
+/// @brief refer to CPU forward -- the BLAS implementation is the same.
+template<typename Dtype>
+void ConvolutionLayer<Dtype>::Forward_gpu(const vector<Blob<Dtype>*>& bottom,
+      const vector<Blob<Dtype>*>& top) {
+
+  const Dtype* weight = this->blobs_[0]->gpu_data();
+  for (int i = 0; i < bottom.size(); ++i) {
+    const Dtype* bottom_data = bottom[i]->gpu_data();
+    Dtype* top_data = top[i]->mutable_gpu_data();
+    for (int n = 0; n < this->num_; ++n) {
+      this->forward_gpu_gemm(bottom_data + bottom[i]->offset(n), weight,
+          top_data + top[i]->offset(n));
+      if (this->bias_term_) {
+        const Dtype* bias = this->blobs_[1]->gpu_data();
+        this->forward_gpu_bias(top_data + top[i]->offset(n), bias);
+      }
+    }
+  }
+}
+
+/// @brief refer to CPU backward -- the BLAS implementation is the same.
+template <typename Dtype>
+void ConvolutionLayer<Dtype>::Backward_gpu(const vector<Blob<Dtype>*>& top,
+      const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom) {
+  const Dtype* weight = this->blobs_[0]->gpu_data();
+  Dtype* weight_diff = this->blobs_[0]->mutable_gpu_diff();
+  if (this->param_propagate_down_[0]) {
+    caffe_gpu_set(this->blobs_[0]->count(), Dtype(0), weight_diff);
+  }
+  if (this->bias_term_ && this->param_propagate_down_[1]) {
+    caffe_gpu_set(this->blobs_[1]->count(), Dtype(0),
+        this->blobs_[1]->mutable_gpu_diff());
+  }
+  for (int i = 0; i < top.size(); ++i) {
+    const Dtype* top_diff = top[i]->gpu_diff();
+    // Bias gradient, if necessary.
+    if (this->bias_term_ && this->param_propagate_down_[1]) {
+      Dtype* bias_diff = this->blobs_[1]->mutable_gpu_diff();
+      for (int n = 0; n < this->num_; ++n) {
+        this->backward_gpu_bias(bias_diff, top_diff + top[i]->offset(n));
+      }
+    }
+    if (this->param_propagate_down_[0] || propagate_down[i]) {
+      const Dtype* bottom_data = bottom[i]->gpu_data();
+      Dtype* bottom_diff = bottom[i]->mutable_gpu_diff();
+      for (int n = 0; n < this->num_; ++n) {
+        // gradient w.r.t. weight. Note that we will accumulate diffs.
+        if (this->param_propagate_down_[0]) {
+          this->weight_gpu_gemm(bottom_data + bottom[i]->offset(n),
+              top_diff + top[i]->offset(n), weight_diff);
+        }
+        // gradient w.r.t. bottom data, if necessary.
+        if (propagate_down[i]) {
+          this->backward_gpu_gemm(top_diff + top[i]->offset(n), weight,
+              bottom_diff + bottom[i]->offset(n));
+        }
+      }
+    }
+  }
+}
+
+#endif // USE_OPENCL
+
+#if defined(CPU_ONLY) && ! defined(USE_OPENCL)
 STUB_GPU(ConvolutionLayer);
 #endif
 
 INSTANTIATE_CLASS(ConvolutionLayer);
+//REGISTER_LAYER_CLASS(Convolution);
```

Comment:
I run into the same issue. But the problem only exists in the Caffe 
version that uses the hand-crafted makefiles and I couldn't fully 
resolve it there. The solution is to use the latest code version and 
build using cmake.

Robert

On 03/30/2015 10:09 AM, jyegerlehner wrote:
>
> In src/caffe/layers/conv_layer.cpp 
> <https://github.com/BVLC/caffe/pull/2195#discussion_r27410109>:
>
> >  STUB_GPU(ConvolutionLayer);
> >  #endif
> >
> >  INSTANTIATE_CLASS(ConvolutionLayer);
> > +//REGISTER_LAYER_CLASS(Convolution);
>
> Commenting out REGISTER_LAYER_CLASS caused tests to fail because.. 
> well.. the layer type isn't registered. Uncommenting this line fixed 
> this problem. Same thing with multiple other layer classes. Please see 
> my commit 
> <https://github.com/jyegerlehner/caffe/commit/6395d002943b8f9f24b3ce9c2035dab83e8a67c5> 
> for all the cases.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/BVLC/caffe/pull/2195/files#r27410109>.
>